ci/scripts/get-bitstream-strategy.sh

```diff
@@ -43,23 +44,61 @@ bitstream_commit=$(ci/bazelisk.sh run //util/py/scripts:get_bitstream_build_id \
   --schema ${REPO_TOP}/rules/scripts/bitstreams_manifest.schema.json \
   --design ${bitstream_design})
 
+bazel_to_path() {
+  # Loop over bazel label dependencies and turn them into paths
+  for dep in $1; do
+    # Ignore everything that does not start with "//"
+    if [[ ! $dep =~ ^// ]]; then
+      continue
+    fi
+    # Replace the : with a / and remove the leading // to get a file name
+    dep=${dep:2}
+    dep=${dep//://}
+    echo $dep
+  done
+}
+
+compute_changed_dep() {
+  for change in $1; do
+    for dep in $2; do
+      if [[ $change == "$dep" ]]; then
+        echo $change
+      fi
+    done
+  done
+}
+
 if [ -z "${bitstream_commit}" ]; then
   echo "Design ${bitstream_design} not found in the cache"
   bitstream_strategy=build
 else
+  # Use bazel to get a list of dependencies for the bitstream target.
+  bazel_labels=$(./bazelisk.sh query "deps(${bitstream_check_target}) + buildfiles(deps(${bitstream_check_target}))")
+  dependencies=$(bazel_to_path "${bazel_labels}")
+
   echo "Checking for changes against pre-built bitstream from ${bitstream_commit}"
   echo "Files changed:"
   git diff --stat --name-only ${bitstream_commit}
   echo
+
+  changed_files=$(git diff --stat --name-only ${bitstream_commit} -- "${excluded_files[@]}")
   echo "Changed files after exclusions applied:"
-  # Use the cached bitstream if no changed files remain.
-  if git diff --exit-code --stat --name-only ${bitstream_commit} -- "${excluded_files[@]}"; then
+  echo "${changed_files}"
+  echo
+
+  # Filter out files not in the dependency list.
+  changed_deps=$(compute_changed_dep "${changed_files}" "${dependencies}")
+  echo "Changed dependencies after exclusions applied:"
+  echo "${changed_deps}"
+  echo
+
+  # Use the cached bitstream if no changed dependencies remain.
+  if [ -z "${changed_deps}" ]; then
```

Comment:
I'm not totally sure this is safe. Are we capturing removed files from the dependency list here? We're still using globs in BUILD files right now, aren't we? So deps changes could get lost, I think.

Reply:
I am not sure I follow what you mean. Can you elaborate?

Reply:
The first step here has changed to get a list of files from the deps of the target at HEAD. However, we don't have a way to detect changes to the contents of that file list (i.e. whether the build graph changed).

Checking the BUILD files for changes might work as a proxy if we explicitly listed each individual file in the filegroup Targets (though there might still be missed files), but we don't do that now. There are just globs all over instead: https://github.com/lowRISC/opentitan/blob/dc22677a68b3f685bad0c2d225b4cf417b8f89c5/hw/ip/aes/BUILD#L9

So changes to the structure of the build graph are not necessarily reflected in diffs of BUILD files.

Reply:
For an example, accidentally removing one of the RTL files captured by the glob would not trigger a bitstream build, would it?

Reply:
That makes sense, you are right that this could be a problem. So are you suggesting that we keep the glob instead of looking at the dependencies on the vivado target then?
We really need to improve the filtering because we are rebuilding way more bitstream than we should so it would be great to merge some improvements quickly.

Reply:
Yeah, I think we should keep the strategy of "always build, except if changes are only in these files that are known not to contribute," at least until we can implement something that accurately tracks differences in the build graph structure.

We could probably check out the cache's chosen commit and compare file lists, but we might want to not do something like this in a rush.

Reply:
If we are not going to merge this as-is, could we at least improve a bit the filtering? We are rebuilding needlessly. For example changes in `rules/` do not affect the bitstream as far as I know? 
I think we could also restructure the azure pipeline using templates so that what concerns the bitstream is isolated in a couple of files. This would avoid retriggering a bitstream build every time we touch the azure pipeline. And it would make the pipeline easier to maintain as well (we have a lot of redundancy). Something inspired by #21714 but less extreme?

Reply:
Yeah, we could definitely improve the filtering, but we can't just exempt all of `rules/`, since `rules/fusesoc.bzl` affects bitstream builds. That directory also has the bitstream cache handling in it. We _could_ reorganize the repo a bit to make writing exclusions easier, though!

At this point, builds can probably be skipped for changes to `azure-pipelines.yml`. The FPGA build is all consolidated in `ci/fpga-template.yml` (plus the scripts it calls).